### PR TITLE
feat (api): create web hook one repository manager

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/workflow/hook.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+		//Perform the request on one off the hooks service
+		if len(srvs) < 1 {
+			return sdk.WrapError(fmt.Errorf("HookRegistration> No hooks service available, please try again"), "Unable to get services dao")
+		}
+
+		var hooksUpdated map[string]sdk.WorkflowNodeHook
+		code, errHooks := services.DoJSONRequest(srvs, http.MethodPost, "/task/bulk", hookToUpdate, &hooksUpdated)
+		if errHooks != nil || code >= 400 {
+			return sdk.WrapError(errHooks, "HookRegistration> Unable to create hooks")
```
Comment (yesnault) [inline engine/api/workflow/hook.go]: add code http in wrap

File: engine/api/workflow/hook.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+		for i := range hooksUpdated {
+			h := hooksUpdated[i]
+			if h.Config["vcsServer"].Value != "" {
+				if err := createVCSConfiguration(db, store, p, &h); err != nil {
+					return sdk.WrapError(err, "HookRegistration> Cannot update vcs configuration")
+				}
+			}
+			if err := UpdateHook(db, &h); err != nil {
+				return sdk.WrapError(errHooks, "HookRegistration> Cannot update hook")
```
Comment (yesnault) [inline engine/api/workflow/hook.go]: err instead of errHooks

File: engine/api/workflow/hook.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		//Push the hook to hooks µService
+		dao := services.Querier(db, store)
+		//Load service "hooks"
+		srvs, err := dao.FindByType("hooks")
+		if err != nil {
+			return sdk.WrapError(err, "HookRegistration> Unable to get services dao")
+		}
+		code, errHooks := services.DoJSONRequest(srvs, http.MethodDelete, fmt.Sprintf("/task/bulk"), hookToDelete, nil)
+		if errHooks != nil || code >= 400 {
+			log.Warning("HookRegistration> Unable to delete old hooks")
```
Comment (yesnault) [inline engine/api/workflow/hook.go]: no wrap here ?

Add http code and errHooks in msg

File: engine/hooks/hooks_handlers.go [hunk 4]
```diff
@@ -216,24 +204,28 @@ func (s *Service) postTaskBulkHandler() api.Handler {
...
 
-			hooks[k] = hook
-
-			//Save the task
-			s.Dao.SaveTask(t)
+func (s *Service) addTask(ctx context.Context, h *sdk.WorkflowNodeHook) error {
+	//Parse the hook as a task
+	t, err := s.hookToTask(h)
+	if err != nil {
+		return sdk.WrapError(err, "Hooks> postTaskBulkHandler> Unable to parse hook")
```
Comment (yesnault) [inline engine/hooks/hooks_handlers.go]: Hooks> addTask>

File: engine/hooks/tasks.go [hunk 5]
```diff
@@ -353,6 +359,10 @@ func executeRepositoryWebHook(t *TaskExecution) (*sdk.WorkflowNodeRunHookEvent,
 		if err := json.Unmarshal(t.WebHook.RequestBody, &pushEvent); err != nil {
 			return nil, sdk.WrapError(err, "Hook> webhookHandler> unable ro read gitlab request: %s", string(t.WebHook.RequestBody))
 		}
+		// Branch deletion
+		if pushEvent.After == "0000000000000000000000000000000000000000" {
```
Comment (yesnault) [inline engine/hooks/tasks.go]: a comment on this value "0000..." could be great

File: engine/hooks/tasks_test.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ func Test_doWebHookExecutionGithub(t *testing.T) {
 		WebHook: &WebHookExecution{
 			RequestBody: []byte(githubPushEvent),
 			RequestHeader: map[string][]string{
-				"X-GitHub-Event": {"push"},
+				"X-Github-Event": {"push"},
```
Comment (yesnault) [inline engine/hooks/tasks_test.go]: you can reuse const GithubHeader

File: engine/vcs/github/client_hook.go [hunk 7]
```diff
@@ -1,13 +1,54 @@
...
+			ContentType: "json",
+		},
+	}
+	b, err := json.Marshal(r)
+	if err != nil {
+		return sdk.WrapError(err, "github.CreateHook > Cannot marshal body %+v", r)
+	}
+	res, err := g.post(url, "application/json", bytes.NewBuffer(b), false)
+	if err != nil {
+		log.Warning("github.CreateHook> Error %s", err)
```
Comment (yesnault) [inline engine/vcs/github/client_hook.go]: wrap?

File: engine/vcs/github/client_hook.go [hunk 8]
```diff
@@ -1,13 +1,54 @@
...
+		return err
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	if res.StatusCode != 201 {
+		err := fmt.Errorf("Unable to create webhook on github. Status code : %d - Body: %s", res.StatusCode, body)
+		log.Warning("github.CreateHook> %s", err)
```
Comment (yesnault) [inline engine/vcs/github/client_hook.go]: wrap?
Group err in one line

File: engine/vcs/github/client_hook.go [hunk 9]
```diff
@@ -1,13 +1,54 @@
...
+		log.Warning("github.CreateHook> Sent data %s", b)
+		return err
+	}
+
+	if err := json.Unmarshal(body, &r); err != nil {
+		log.Warning("github.CreateHook> Cannot unmarshal response")
```
Comment (yesnault) [inline engine/vcs/github/client_hook.go]: wrap?

File: engine/vcs/github/http.go [hunk 10]
```diff
@@ -201,3 +201,41 @@ func (c *githubClient) get(path string, opts ...getArgFunc) (int, []byte, http.H
...
+		return err
+	}
+
+	req.Header.Set("User-Agent", "CDS-gh_client_id="+c.ClientID)
+	req.Header.Add("Authorization", fmt.Sprintf("token %s", c.OAuthToken))
+	log.Debug("Github API>> Request URL %s", req.URL.String())
+
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return err
```
Comment (yesnault) [inline engine/vcs/github/http.go]: wrap

File: engine/api/workflow/hook.go (hunk 3 above)
Comment (sguiheux) [reply engine/api/workflow/hook.go]: I do not want to rollback because of an error on repo manager size.

File: engine/vcs/github/client_hook.go [hunk 11]
```diff
@@ -1,13 +1,49 @@
...
+		return sdk.WrapError(err, "github.CreateHook > Cannot marshal body %+v", r)
+	}
+	res, err := g.post(url, "application/json", bytes.NewBuffer(b), false)
+	if err != nil {
+		return sdk.WrapError(err, "github.CreateHook")
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
```
Comment (yesnault) [inline engine/vcs/github/client_hook.go]: Wrap

File: engine/vcs/gitlab/client_hook.go [hunk 12]
```diff
@@ -19,13 +20,19 @@ func (g *gitlabClient) UpdateHook(repo, id string, hook sdk.VCSHook) error {
...
 
-	url, err := buildGitlabURL(hook.URL)
-	if err != nil {
-		return err
+	var url string
+	if !hook.Workflow {
+		var err error
+		url, err = buildGitlabURL(hook.URL)
+		if err != nil {
+			return err
```
Comment (yesnault) [inline engine/vcs/gitlab/client_hook.go]: wrap

File: engine/vcs/vcs_handlers.go [hunk 13]
```diff
@@ -639,7 +631,10 @@ func (s *Service) postHookHandler() api.Handler {
 			return sdk.WrapError(err, "VCS> postHookHandler> Unable to read body")
 		}
 
-		return client.CreateHook(fmt.Sprintf("%s/%s", owner, repo), body)
+		if err := client.CreateHook(fmt.Sprintf("%s/%s", owner, repo), &body); err != nil {
+			return err
```
Comment (yesnault) [inline engine/vcs/vcs_handlers.go]: wrap

File: engine/vcs/gitlab/client_hook.go [hunk 14]
```diff
@@ -37,37 +44,56 @@ func (c *gitlabClient) CreateHook(repo string, hook sdk.VCSHook) error {
...
 
 //DeleteHook disables the defaut HTTP POST Hook in Gitlab
 func (c *gitlabClient) DeleteHook(repo string, hook sdk.VCSHook) error {
+	if !hook.Workflow {
+		var url string
+		if !hook.Workflow {
+			var err error
+			url, err = buildGitlabURL(hook.URL)
+			if err != nil {
+				return err
```
Comment (yesnault) [inline engine/vcs/gitlab/client_hook.go]: wrap

File: engine/vcs/gitlab/client_hook.go [hunk 15]
```diff
@@ -37,37 +44,56 @@ func (c *gitlabClient) CreateHook(repo string, hook sdk.VCSHook) error {
 	}
 
 	log.Debug("GitlabClient.CreateHook: %s %s\n", repo, *opt.URL)
-	if _, _, err := c.client.Projects.AddProjectHook(repo, &opt); err != nil {
+	ph, resp, err := c.client.Projects.AddProjectHook(repo, &opt)
+	if err != nil {
 		return err
```
Comment (yesnault) [inline engine/vcs/gitlab/client_hook.go]: wrap

File: engine/vcs/gitlab/client_hook.go [hunk 16]
```diff
@@ -37,37 +44,56 @@ func (c *gitlabClient) CreateHook(repo string, hook sdk.VCSHook) error {
...
-	}
 
-	return fmt.Errorf("not found")
+		log.Debug("GitlabClient.DeleteHook: Got '%s'", url)
+		log.Debug("GitlabClient.DeleteHook: Want '%s'", url)
+		for _, h := range hooks {
+			log.Debug("GitlabClient.DeleteHook: Found '%s'", h.URL)
+			if h.URL == url {
+				_, err = c.client.Projects.DeleteProjectHook(repo, h.ID)
+				return err
```
Comment (yesnault) [inline engine/vcs/gitlab/client_hook.go]: wrap

File: engine/api/workflow.go [hunk 17]
```diff
@@ -127,23 +124,8 @@ func (api *API) postWorkflowHandler() Handler {
...
-		hooks := wf.GetHooks()
-		if len(hooks) > 0 {
-			if len(srvs) < 1 {
-				return sdk.WrapError(fmt.Errorf("postWorkflowHandler> No hooks service available, please try again"), "Unable to get services dao")
-			}
-			if _, errHooks := services.DoJSONRequest(srvs, http.MethodPost, "/task/bulk", hooks, nil); errHooks != nil {
-				return sdk.WrapError(errHooks, "postWorkflowHandler> Unable to create hooks")
-			}
+		if err := workflow.HookRegistration(tx, api.Cache, nil, wf, p); err != nil {
+			return sdk.WrapError(err, "putWorkflowHandler")
```
Comment (fsamin) [inline engine/api/workflow.go]: postWorkflowHandler

File: engine/hooks/scheduler.go [hunk 18]
```diff
@@ -122,17 +121,20 @@ func (s *Service) dequeueTaskExecutions(c context.Context) error {
 		// Load the task execution
 		var t = TaskExecution{}
 		if !s.Cache.Get(taskKey, &t) {
+			log.Debug("TaskExecution not found: %s", taskKey)
```
Comment (fsamin) [inline engine/hooks/scheduler.go]: ploup

File: engine/hooks/scheduler.go [hunk 19]
```diff
@@ -122,17 +121,20 @@ func (s *Service) dequeueTaskExecutions(c context.Context) error {
...
 			continue
 		}
 		t.ProcessingTimestamp = time.Now().UnixNano()
 		s.Dao.SaveTaskExecution(&t)
 
 		task := s.Dao.FindTask(t.UUID)
+		log.Debug("FindTask : %s", task.UUID)
```
Comment (fsamin) [inline engine/hooks/scheduler.go]: ploup

File: engine/hooks/scheduler.go [hunk 20]
```diff
@@ -122,17 +121,20 @@ func (s *Service) dequeueTaskExecutions(c context.Context) error {
...
 		if task == nil {
 			log.Error("Hooks> dequeueTaskExecutions failed: Task not found")
 			t.LastError = "Internal Error: Task not found"
 			t.NbErrors++
 		} else if task.Stopped {
+			log.Debug("Task Stooped : %s", task.UUID)
```
Comment (fsamin) [inline engine/hooks/scheduler.go]: ploup

File: engine/hooks/tasks.go [hunk 21]
```diff
@@ -333,12 +333,16 @@ func executeRepositoryWebHook(t *TaskExecution) (*sdk.WorkflowNodeRunHookEvent,
 	}
 
 	payload := make(map[string]interface{})
+	log.Debug("Header: %s", getRepositoryHeader(t.WebHook))
```
Comment (fsamin) [inline engine/hooks/tasks.go]: ploup

File: engine/api/workflow/hook.go [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+			}
+			if err := UpdateHook(db, &h); err != nil {
+				return sdk.WrapError(err, "HookRegistration> Cannot update hook")
+			}
+		}
+	}
+
+	if len(hookToDelete) > 0 {
+		if err := deleteHookConfiguration(db, store, p, hookToDelete); err != nil {
+
```
Comment (bnjjj) [inline engine/api/workflow/hook.go]: missing something here

File: engine/api/workflow/hook.go [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+					return sdk.WrapError(errclient, "deleteVCSConfiguration> Cannot get vcs client")
+				}
+				vcsHook := sdk.VCSHook{
+					Method:   "POST",
+					URL:      h.Config["webHookURL"].Value,
+					Workflow: true,
+					ID:       h.Config["webHookID"].Value,
+				}
+				if err := client.DeleteHook(h.Config["repoFullName"].Value, vcsHook); err != nil {
+					return sdk.WrapError(err, "deleteVCSConfiguration> Cannot delete hook on repository")
```
Comment (bnjjj) [inline engine/api/workflow/hook.go]: deleteHookConfiguration>

File: engine/api/workflow/hook.go [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+func deleteHookConfiguration(db gorp.SqlExecutor, store cache.Store, p *sdk.Project, hookToDelete map[string]sdk.WorkflowNodeHook) error {
+	// Delete from vcs configuration if needed
+	for _, h := range hookToDelete {
+		if h.WorkflowHookModel.Name == RepositoryWebHookModel.Name {
+			// Call VCS to know if repository allows webhook and get the configuration fields
+			projectVCSServer := repositoriesmanager.GetProjectVCSServer(p, h.Config["vcsServer"].Value)
+			if projectVCSServer != nil {
+				client, errclient := repositoriesmanager.AuthorizedClient(db, store, projectVCSServer)
+				if errclient != nil {
+					return sdk.WrapError(errclient, "deleteVCSConfiguration> Cannot get vcs client")
```
Comment (bnjjj) [inline engine/api/workflow/hook.go]: deleteHookConfiguration>

File: engine/api/workflow/hook.go [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+	if errHooks != nil || code >= 400 {
+		log.Warning("HookRegistration> Unable to delete old hooks [%d]: %s", code, errHooks)
+	}
+	return nil
+}
+
+func createVCSConfiguration(db gorp.SqlExecutor, store cache.Store, p *sdk.Project, h *sdk.WorkflowNodeHook) error {
+	// Call VCS to know if repository allows webhook and get the configuration fields
+	projectVCSServer := repositoriesmanager.GetProjectVCSServer(p, h.Config["vcsServer"].Value)
+	if projectVCSServer != nil {
```
Comment (bnjjj) [inline engine/api/workflow/hook.go]: maybe better
```
if projectVCSServer == nil {
return nil
}
```

File: engine/vcs/gitlab/client_hook.go [hunk 26]
```diff
@@ -37,37 +44,56 @@ func (c *gitlabClient) CreateHook(repo string, hook sdk.VCSHook) error {
...
-		if h.URL == url {
-			_, err = c.client.Projects.DeleteProjectHook(repo, h.ID)
-			return err
+		log.Debug("GitlabClient.DeleteHook: Got '%s'", url)
+		log.Debug("GitlabClient.DeleteHook: Want '%s'", url)
+		for _, h := range hooks {
+			log.Debug("GitlabClient.DeleteHook: Found '%s'", h.URL)
+			if h.URL == url {
+				_, err = c.client.Projects.DeleteProjectHook(repo, h.ID)
+				return sdk.WrapError(err, "github.DeleteHook> DeleteProjectHook")
```
Comment (bnjjj) [inline engine/vcs/gitlab/client_hook.go]: missing if err != nil ?

File: engine/vcs/gitlab/client_hook.go [hunk 27]
```diff
@@ -37,37 +44,56 @@ func (c *gitlabClient) CreateHook(repo string, hook sdk.VCSHook) error {
...
+			}
 		}
+		return fmt.Errorf("not found")
```
Comment (bnjjj) [inline engine/vcs/gitlab/client_hook.go]: `github.DeleteHook>`

File: engine/vcs/gitlab/client_hook.go (hunk 26 above)
Comment (sguiheux) [reply engine/vcs/gitlab/client_hook.go]: wrapError of nil == nil

**Pull request merged**